File: sdk/cpp/tests/test_vector_add.cpp
// Copyright 2026 Ojima Abraham
// SPDX-License-Identifier: Apache-2.0

/// End-to-end test for the WAVE C/C++ SDK.

#include "wave/wave.h"

#include <cassert>
#include <cmath>
#include <cstdio>

static void test_device_detection() {
    wave_device_t *dev = wave_detect_device();
    assert(dev != nullptr);

    wave_vendor_t vendor = wave_device_vendor(dev);
    assert(vendor >= WAVE_VENDOR_APPLE && vendor <= WAVE_VENDOR_EMULATOR);

    const char *name = wave_device_name(dev);
    assert(name != nullptr);
    assert(name[0] != '\0');

    printf("Device: %s\n", name);
    wave_free_device(dev);
}

static void test_buffer_f32() {
    float data[] = {1.0f, 2.0f, 3.0f, 4.0f};
    wave_buffer_t *buf = wave_create_buffer_f32(data, 4);
    assert(buf != nullptr);

    float out[4] = {0};
    int ret = wave_read_buffer_f32(buf, out, 4);
    assert(ret == 0);
    assert(out[0] == 1.0f);
    assert(out[1] == 2.0f);
    assert(out[2] == 3.0f);
    assert(out[3] == 4.0f);

    wave_free_buffer(buf);
}

static void test_zeros_f32() {
    wave_buffer_t *buf = wave_create_zeros_f32(8);
    assert(buf != nullptr);

    float out[8] = {0};
    int ret = wave_read_buffer_f32(buf, out, 8);
    assert(ret == 0);
    for (int i = 0; i < 8; i++) {
        assert(out[i] == 0.0f);
    }

    wave_free_buffer(buf);
}

int main() {
    printf("Running WAVE C/C++ SDK tests...\n");

    test_device_detection();
    printf("  [PASS] device detection\n");

    test_buffer_f32();
    printf("  [PASS] buffer f32 roundtrip\n");

    test_zeros_f32();
    printf("  [PASS] zeros f32\n");

    printf("All C/C++ SDK tests passed!\n");
    return 0;
}

File: sdk/js/tests/test_vector_add.ts
// Copyright 2026 Ojima Abraham
// SPDX-License-Identifier: Apache-2.0

/// End-to-end test for the WAVE JavaScript/TypeScript SDK.

import { device, array, zeros, WaveArray } from "../src/index";

async function testDevice(): Promise<void> {
  const dev = await device();
  console.assert(dev.vendor !== undefined, "vendor should be defined");
  console.assert(dev.name.length > 0, "name should be non-empty");
  console.log(`  Device: ${dev.name}`);
}

function testArray(): void {
  const a = array([1, 2, 3, 4]);
  console.assert(a.length === 4, "length should be 4");
  console.assert(
    JSON.stringify(a.toArray()) === JSON.stringify([1, 2, 3, 4]),
    "toArray should match",
  );

  const z = zeros(5);
  console.assert(z.length === 5, "zeros length should be 5");
  console.assert(
    z.toArray().every((v) => v === 0),
    "zeros should all be 0",
  );
}

function testBufferRoundtrip(): void {
  const a = array([1.5, 2.5, 3.5]);
  const buf = a.toBuffer();
  const b = WaveArray.fromBuffer(buf, 3);
  const diff = a
    .toArray()
    .map((v, i) => Math.abs(v - b.toArray()[i]))
    .reduce((s, d) => s + d, 0);
  console.assert(diff < 0.001, "buffer roundtrip should preserve values");
}

async function main(): Promise<void> {
  console.log("Running WAVE JS/TS SDK tests...");

  await testDevice();
  console.log("  [PASS] device detection");

  testArray();
  console.log("  [PASS] array creation");

  testBufferRoundtrip();
  console.log("  [PASS] buffer roundtrip");

  console.log("All JS/TS SDK tests passed!");
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
